File: aizu online judge/練習/3章 初等的整列/python/Insertion sort.py
"""
https://judge.u-aizu.ac.jp/onlinejudge/description.jsp?id=ALDS1_1_A&lang=ja

解答例を再現
"""

def trace(a, n):
    print(" ".join(list(map(str, a))))

def insertion_sort(a, n):
    for i in range(1, n):
        v = a[i]
        j = i - 1
        while j >= 0 and a[j] > v:
            a[j + 1] = a[j]
            j -= 1
            print("v: {}".format(v))
            print("a: {}".format(a))
        a[j + 1] = v
        trace(a, n)

n = int(input())
a = list(map(int, input().split()))
trace(a, n)
insertion_sort(a, n)

File: aizu online judge/練習/3章 初等的整列/C++/insertion_sort.cpp
// https://judge.u-aizu.ac.jp/onlinejudge/description.jsp?id=ALDS1_1_A&lang=ja

#include <iostream>
using namespace std;

void trace(int A[],int N) {
  for (int i = 0; i < N; i++) {
    if (i > 0) cout << " ";
    cout << A[i];
  }
  cout << endl;
}

void insertionSort(int A[], int N) {
  int v;
  for (int i = 1; i < N; i++) {
    v = A[i];
    int j = i - 1;
    while (j >= 0 && A[j] > v) {
      A[j + 1] = A[j];
      j--;
    }
    A[j + 1] = v;
    trace(A, N);
  }
}

int main() {
  int N;
  cin >> N;
  int A[N];

  for (int i = 0; i < N; i++) {
    cin >> A[i];
  }
  trace(A, N);
  insertionSort(A, N);
  return 0;
}